File: 572-is-subtree/solution.go
package main
import(
	"fmt"
	"gopractice/datastructure/Tree"
)
type TreeNode = Tree.TreeNode
func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
	if root == nil && subRoot == nil {
		return true
	} else if root == nil || subRoot == nil {
		return false
	} else if root.Val != subRoot.Val {
		return isSubtree(root.Left, subRoot) || isSubtree(root.Right, subRoot)
	} else {
		return  (isSameTree(root.Left, subRoot.Left) && isSameTree(root.Right, subRoot.Right)) ||  isSubtree(root.Left, subRoot) || isSubtree(root.Right, subRoot)
	}
}

func isSameTree(p *TreeNode, q *TreeNode) bool {
    if p == nil && q == nil {
		return true
	} else if p == nil || q == nil {
		return false
	} else if p.Val != q.Val {
		return false
	} else {
		return isSameTree(p.Left, q.Left) && isSameTree(p.Right, q.Right)
	}
}

func main(){
	root := []any {3,4,5,1,2,nil,nil,nil,nil,0}
	subRoot := []any {4,1,2}
	rootNode := Tree.BuildTree(root)
	subRootNode := Tree.BuildTree(subRoot)
	fmt.Println("Ans = ", isSubtree(rootNode, subRootNode))
}

File: 572-is-subtree/solution.js
const TreeNode = require('../datastructure/Tree/TreeNode');
const buildLevelOrderTree_v2 = require('../datastructure/Tree/TreeDebugger_v2')
/**
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 */
var isSubtree = function(root, subRoot) {
    if(root === null && subRoot === null)
        return true
    else if(root === null || subRoot === null)
        return false
    else if(root.val !== subRoot.val)
        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot)
    else
        return isSametree(root, subRoot) || isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot)
};
var isSametree = function(root, subRoot) {
    if(root === null && subRoot === null)
        return true
    else if(root === null || subRoot === null)
        return false
    else if(root.val !== subRoot.val)
        return false
    else
        return isSametree(root.left, subRoot.left) && isSametree(root.right, subRoot.right)
}
if(require.main === module){
    const levelOrder = [3,4,5,1,2]
    const root = buildLevelOrderTree_v2(levelOrder)
    const subTree = [4,1,2]
    const subRoot = buildLevelOrderTree_v2(subTree)
    console.log("Ans = ", isSubtree(root, subRoot))
}
//寫第三次還是沒有寫出來